src/desktop.c

```diff
@@ -293,92 +293,74 @@ desktop_focus_topmost_mapped_view(struct server *server)
 	desktop_move_to_front(view);
 }
 
-struct view *
-desktop_node_and_view_at(struct server *server, double lx, double ly,
-		struct wlr_scene_node **scene_node, double *sx, double *sy,
-		enum ssd_part_type *view_area)
+/* TODO: make this less big and scary */
+struct cursor_context
+get_cursor_context(struct server *server)
 {
+	struct cursor_context ret = {.type = LAB_SSD_NONE};
+	struct wlr_cursor *cursor = server->seat.cursor;
 	struct wlr_scene_node *node =
-		wlr_scene_node_at(&server->scene->tree.node, lx, ly, sx, sy);
+		wlr_scene_node_at(&server->scene->tree.node,
+			cursor->x, cursor->y, &ret.sx, &ret.sy);
 
-	*scene_node = node;
+	ret.node = node;
 	if (!node) {
-		*view_area = LAB_SSD_ROOT;
-		return NULL;
+		ret.type = LAB_SSD_ROOT;
+		return ret;
 	}
 	if (node->type == WLR_SCENE_NODE_BUFFER) {
 		struct wlr_surface *surface = lab_wlr_surface_from_node(node);
```

Comment:
Could we already set the surface here? AFAIR that whole SSD enum check of what is a surface is technical debt from before scene-graph.

Reply:
I tried to stay as close to the existing logic as possible here just for fear of not fully understanding the implications, and breaking something. So that meant setting `ret.surface` only in the three cases covered by the old `is_surface()` check:

- `LAB_SSD_CLIENT`
- `LAB_SSD_LAYER_SURFACE`
- `LAB_SSD_UNMANAGED`

I expect there are more cleanups possible here, but I'd rather leave them for a future change.